sklearn/neighbors/quad_tree.pyx

```diff
@@ -0,0 +1,592 @@
+                         ) nogil:
+        """Initialize the root node with the given space boundaries"""
+        cdef:
+            int i
+            DTYPE_t width
+            Cell* root = &self.cells[0]
+
+        self._init_cell(root, -1, 0)
+        for i in range(self.n_dimensions):
+            root.min_bounds[i] = min_bounds[i]
+            root.max_bounds[i] = max_bounds[i]
+            root.center[i] = (max_bounds[i] + min_bounds[i]) / 2.
+            width = max_bounds[i] - min_bounds[i]
+            root.max_width = max(root.max_width, width*width)
+        root.cell_id = 0
+
+        self.cell_count += 1
+
+    def plot_tree(self):
+        """Plot the tree with cell boundaries and the points inserted in it."""
+        self.check_coherence()
+        import matplotlib.pyplot as plt
+
+        plt.figure()
+        for c in self.cells[:self.cell_count]:
+            if not c.is_leaf:
+                # Plot the cell division if the cell is an inner cell
+                plt.vlines(c.center[0], c.min_bounds[1], c.max_bounds[1])
+                plt.hlines(c.center[1], c.min_bounds[0], c.max_bounds[0])
+            else:
+                # If the cell is a leaf, display the point contained in it.
+                plt.scatter(c.barycenter[0], c.barycenter[1], c='b', marker='.')
+
+        # Print bounding box of the Tree
+        root = self.cells[0]
+        plt.vlines([root.min_bounds[0], root.max_bounds[0]], root.min_bounds[1], root.max_bounds[1])
+        plt.hlines([root.min_bounds[1], root.max_bounds[1]], root.min_bounds[0], root.max_bounds[0])
+        plt.show()
+
+    cdef int check_point_in_cell(self, DTYPE_t[3] point, Cell* cell
+                                  ) nogil except -1:
+        """Check that the given point is in the cell boundaries."""
+
+        if self.verbose >= 10:
+            printf("[QuadTree] Checking point (%f, %f, %f) in cell %li "
+                    "([%f/%f, %f/%f, %f/%f], size %li)\n",
+                    point[0], point[1], point[2], cell.cell_id,
+                    cell.min_bounds[0], cell.max_bounds[0], cell.min_bounds[1],
+                    cell.max_bounds[1], cell.min_bounds[2], cell.max_bounds[2],
+                    cell.cumulative_size)
+
+        for i in range(self.n_dimensions):
+            if (cell.min_bounds[i] > point[i] or
+                    cell.max_bounds[i] <= point[i]):
+                with gil:
+                    msg = "[QuadTree] InsertionError: point out of cell boundary.\n"
+                    msg += "Axis %li: cell [%f, %f]; point %f\n"
+
+                    msg %= i, cell.min_bounds[i],  cell.max_bounds[i], point[i]
+                    raise ValueError(msg)
+
+    def check_coherence(self):
+        """Check the coherence of the cells of the tree.
+
+        Check that the info stored in each cell are compatible with the info
+        stored in descendent and sibling cells. Raise a RuntimeError if this
+        fails.
```

Comment:
I would rather raise a `ValueError` I think here (the "value" being the tree itself which is invalid).